### ISSUE TITLE
Validation: Implement Walk-Forward Optimization to Detect Overfitting

### ISSUE THREAD

**Issue by pselamy** [tradestreamhq/tradestream#1560]
## Overview

Implement walk-forward optimization/validation to detect overfitting before strategies are deployed. This is the **single most important technical change** to validate that discovered strategies generalize to unseen data.

---

## The Problem: Overfitting

From #1555 (Strategic Assessment):

> "Quantopian's 888-strategy study found that Sharpe ratios from backtests had **almost zero predictive power** for live returns."

Our current GA pipeline:
1. Optimizes on ALL historical data
2. Reports in-sample metrics
3. Has no mechanism to detect overfitting

**This means we don't know if our 40M+ discoveries are real edges or noise.**

---

## The Solution: Walk-Forward Validation

```
┌─────────────────────────────────────────────────────────────────────┐
│  CURRENT FLOW (OVERFITTING RISK)                                   │
├─────────────────────────────────────────────────────────────────────┤
│  [───────── ALL DATA ──────────]                                   │
│                 ↓                                                   │
│              GA Optimizes                                           │
│                 ↓                                                   │
│         "Best" Strategy Found                                       │
│                 ↓                                                   │
│         (probably overfit)                                          │
└─────────────────────────────────────────────────────────────────────┘

┌─────────────────────────────────────────────────────────────────────┐
│  NEW FLOW (WITH WALK-FORWARD)                                      │
├─────────────────────────────────────────────────────────────────────┤
│  [── TRAIN ──][─TEST─][── TRAIN ──][─TEST─][── TRAIN ──][─TEST─]   │
│       ↓          ↓          ↓          ↓          ↓          ↓     │
│    GA Opt    Validate    GA Opt    Validate    GA Opt    Validate  │
│       ↓          ↓          ↓          ↓          ↓          ↓     │
│              Aggregate out-of-sample performance                    │
│                              ↓                                      │
│                  Strategy validated or rejected                     │
└─────────────────────────────────────────────────────────────────────┘
```

---

## Implementation

Walk-forward validation will be implemented **within the Backtesting Service** (#1558), leveraging ta4j 0.22's walk-forward helpers and the new `BacktestExecutor`.

### Phase 1: Data Splitting Infrastructure

```java
public class WalkForwardConfig {
    private final int trainWindowBars;    // e.g., 6480 bars (9 months @ 1min)
    private final int testWindowBars;     // e.g., 2160 bars (3 months @ 1min)
    private final int stepBars;           // e.g., 2160 bars (3 months)
    private final int minWindows;         // e.g., 3 windows minimum
}

public class WalkForwardSplitter {
    public List<TrainTestSplit> createSplits(BarSeries series, WalkForwardConfig config) {
        List<TrainTestSplit> splits = new ArrayList<>();
        int start = 0;
        while (start + config.trainWindowBars + config.testWindowBars <= series.getBarCount()) {
            splits.add(new TrainTestSplit(
                series.getSubSeries(start, start + config.trainWindowBars),
                series.getSubSeries(start + config.trainWindowBars, 
                                   start + config.trainWindowBars + config.testWindowBars)
            ));
            start += config.stepBars;
        }
        return splits;
    }
}
```

### Phase 2: gRPC Endpoint in Backtesting Service

Add walk-forward endpoint to #1558's gRPC service:

```protobuf
// Add to backtesting_service.proto
message WalkForwardRequest {
  string strategy_name = 1;
  repeated Candle candles = 2;
  WalkForwardConfig config = 3;
  Strategy strategy_params = 4;  // Optional: validate specific params
}

message WalkForwardResult {
  double in_sample_sharpe = 1;
  double out_of_sample_sharpe = 2;
  double sharpe_degradation = 3;
  int32 windows_count = 4;
  ValidationStatus status = 5;
  repeated WindowResult window_results = 6;
}

service BacktestService {
  // ... existing methods from #1558 ...
  rpc RunWalkForwardValidation(WalkForwardRequest) returns (WalkForwardResult);
}
```

### Phase 3: Validation Gate

```java
public class StrategyValidator {
    private static final double MIN_OOS_SHARPE = 0.5;
    private static final double MAX_SHARPE_DEGRADATION = 0.5;  // OOS must be >= 50% of IS
    
    public ValidationResult validate(WalkForwardResult wfResult) {
        if (wfResult.getOutOfSampleSharpe() < MIN_OOS_SHARPE) {
            return ValidationResult.rejected("OOS Sharpe too low: " + wfResult.getOutOfSampleSharpe());
        }
        
        if (wfResult.getSharpeDegradation() > MAX_SHARPE_DEGRADATION) {
            return ValidationResult.rejected("Sharpe degradation too high: " + 
                String.format("%.1f%%", wfResult.getSharpeDegradation() * 100));
        }
        
        return ValidationResult.approved(wfResult);
    }
}
```

---

## Database Schema Changes

```sql
-- Add OOS metrics to discovered_strategies table
ALTER TABLE discovered_strategies ADD COLUMN oos_sharpe DOUBLE PRECISION;
ALTER TABLE discovered_strategies ADD COLUMN oos_return DOUBLE PRECISION;
ALTER TABLE discovered_strategies ADD COLUMN sharpe_degradation DOUBLE PRECISION;
ALTER TABLE discovered_strategies ADD COLUMN validation_status VARCHAR(20);
ALTER TABLE discovered_strategies ADD COLUMN wf_windows_count INTEGER;

-- Index for finding validated strategies
CREATE INDEX idx_validated_strategies 
ON discovered_strategies(validation_status, oos_sharpe DESC)
WHERE validation_status = 'APPROVED';
```

---

## Metrics to Track

| Metric | Formula | Healthy Range |
|--------|---------|---------------|
| **OOS Sharpe** | Mean Sharpe across test windows | > 0.5 |
| **Sharpe Degradation** | 1 - (OOS Sharpe / IS Sharpe) | < 50% |
| **OOS Consistency** | StdDev of Sharpe across windows | < 0.5 |
| **Approval Rate** | Approved / Total Discoveries | 5-20% |

**If approval rate is too high (>50%), validation is too lenient.**
**If approval rate is too low (<1%), validation is too strict or strategies are garbage.**

---

## Acceptance Criteria

- [ ] `WalkForwardSplitter` creates train/test windows correctly
- [ ] gRPC endpoint added to Backtesting Service (#1558)
- [ ] GA runs on training data only, never sees test data
- [ ] OOS metrics calculated and stored in database
- [ ] Validation gate rejects overfit strategies
- [ ] Approval rate tracked in monitoring dashboard
- [ ] At least 3 walk-forward windows used per evaluation

---

## Dependencies

- #1558 (Backtesting Service) - Walk-forward runs in the new service

## Blocks

- #1485 (Signal Service) - Only validated strategies should generate signals
- #1487 (Sprint Plan) - Adds validation gate before launch

---

## Research References

- [Walk-Forward Optimization Explained](https://www.quantifiedstrategies.com/walk-forward-optimization/)
- [Ta4j Walk-Forward Examples](https://ta4j.github.io/ta4j-wiki/Backtesting.html)
- [Overfitting in Trading](https://algotrading101.com/learn/what-is-overfitting-in-trading/)

## Labels
`priority-critical`, `enhancement`, `backtesting`
